src/main/java/org/zalando/nakadi/filters/LoggingFilter.java

```diff
@@ -34,47 +38,123 @@ public LoggingFilter(final NakadiKpiPublisher nakadiKpiPublisher,
         this.accessLogEventType = accessLogEventType;
     }
 
+    private class RequestLogInfo {
+
+        private String userAgent;
+        private String user;
+        private String method;
+        private String path;
+        private String query;
+        private String contentEncoding;
+        private Long contentLength;
+        private String acceptEncoding;
+        private Long requestTime;
+
+        private RequestLogInfo(final HttpServletRequest request, final long requestTime) {
+            this.userAgent = Optional.ofNullable(request.getHeader("User-Agent")).orElse("-");
+            this.user = Optional.ofNullable(request.getUserPrincipal()).map(Principal::getName).orElse("-");
+            this.method = request.getMethod();
+            this.path = request.getRequestURI();
+            this.query = Optional.ofNullable(request.getQueryString()).map(q -> "?" + q).orElse("");
+            this.contentEncoding = Optional.ofNullable(request.getHeader(HttpHeaders.CONTENT_ENCODING))
+                .orElse("-");
+            this.acceptEncoding = Optional.ofNullable(request.getHeader(HttpHeaders.ACCEPT_ENCODING))
+                .orElse("-");
+            this.contentLength = request.getContentLengthLong() == -1 ? 0 : request.getContentLengthLong();
+            this.requestTime = requestTime;
+        }
+    }
+
+    private class AsyncRequestListener implements AsyncListener {
+        private final HttpServletResponse response;
+        private final String flowId;
+        private final RequestLogInfo requestLogInfo;
+
+        private AsyncRequestListener(final HttpServletRequest request, final HttpServletResponse response,
+                                    final long startTime, final String flowId) {
+            this.response = response;
+            this.flowId = flowId;
+
+            this.requestLogInfo = new RequestLogInfo(request, startTime);
+            ACCESS_LOGGER.info("{} \"{}{}\" \"{}\" \"{}\" {} {}ms \"{}\" \"{}\" {}B",
+                    requestLogInfo.method,
+                    requestLogInfo.path,
+                    requestLogInfo.query,
+                    requestLogInfo.userAgent,
+                    requestLogInfo.user,
+                    HttpStatus.PROCESSING.value(),
+                    0,
+                    requestLogInfo.contentEncoding,
+                    requestLogInfo.acceptEncoding,
+                    requestLogInfo.contentLength);
+        }
+
+        @Override
+        public void onComplete(final AsyncEvent event) {
+            FlowIdUtils.push(this.flowId);
```

Comment:
you also need to clean up after that, because the flowid will be left on the thread which calls AsyncRequestListener